### PR TITLE
feat: add datasource config in import module

### LINKED ISSUE #82: support multiple types of data sources in the studio
**Is your feature request related to a problem? Please describe.**
When i want to load data by studio, i can only use .csv files. 
If I have some other types of data sources, such as hive and spark. It will be a very complicated thing.

**Describe the solution you'd like**
Maybe we can set a "data source management" in the studio, so that we can do some config and load data visually. 

**Describe alternatives you've considered**
manually config in the terminal.

**Additional context**

### LINKED ISSUE #83: support clear logs of data import
**Is your feature request related to a problem? Please describe.**
When i load data, i can just see some errors occur. But i don't know what happens.
Last time i found some errors occurred, i found nothing in the studio. Finally i found that the storaged was down.

**Describe the solution you'd like**
We can consider to put some clear logs in the studio, especially when we load data.

**Describe alternatives you've considered**
in the terminal

**Additional context**
image

### LINKED ISSUE #188: The advice about upload
Uploading a file with the same name will directly overwrite the old file. It is recommended to add a reminder or rename the file automatically. 
In addition, it is recommended to add a column showing the upload time.

### LINKED ISSUE #512: [多数据源导入] 本地文件 弹框路径刷新引发页面崩溃
刷新页面崩溃

### LINKED ISSUE #513: [多数据源导入] 生成导入任务报错，提示id没有索引
image

image

### LINKED ISSUE #516: suggestions on data source support for data import
## i18n suggestions

- "Cloud storage" --> "Object Storage"
- "S3 Platform" --> "S3 compatible Service Provider"

### PR DESCRIPTION
**Pull request by hetao92**
## What type of PR is this?
- [x] bug
- [x] feature
- [ ] enhancement

#### Issue(s) number: 
- fix #513 
- fix #512 
- fix #516

### CODE REVIEW

File: app/config/service.ts [hunk 1]
```diff
@@ -90,6 +90,32 @@ const service = {
...
+  },
+  updateDatasource: (params, config?) => {
+    const { id, ...restParams } = params;
+    return post(`/api/datasources/${id}`)(restParams, config);
+  },
+  deleteDatasource: (id: number, config?) => {
+    return _delete(`/api/datasources/${id}`)(undefined, config);
+  },
+  batchDeleteDatasource: (payload, config?) => {
+    return _delete(`/api/datasources`)(undefined, { data: payload });
```
Comment (jerryliang64) [inline app/config/service.ts:109]: recommand: undefined => {}?

File: app/interfaces/import.ts [hunk 2]
```diff
@@ -47,14 +49,40 @@ export interface IImportFile {
...
+  /** remote path */
+  path?: string;
+}
+
+export interface IS3Config {
+  region: string;
+  endpoint: string;
+  accessKey: string;
+  accessSecret: string;
+  bucket?: string;
```
Comment (jerryliang64) [inline app/interfaces/import.ts]: this is required now?

File: app/interfaces/import.ts [hunk 3]
```diff
@@ -47,14 +49,40 @@ export interface IImportFile {
...
+  token?: string;
+  key: string;
```
Comment (jerryliang64) [inline app/interfaces/import.ts:67]: what are the token and key?

File: app/interfaces/import.ts [hunk 4]
```diff
@@ -47,14 +49,40 @@ export interface IImportFile {
...
+export interface ISftpConfig {
+  host: string;
+  port: number;
+  username: string;
+  password: string;
+  path: string;
+  keyFile?: string;
+  keyData?: string;
+  passPhrase?: string;
+}
```
Comment (jerryliang64) [inline app/interfaces/import.ts:79]: what are they used for?

File: app/pages/Import/DatasourceList/DatasourceConfig/PlatformConfig/S3ConfigForm.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+      setFlag(true);
+    }
+  };
+  return (
+    <FormItem noStyle>
+      <FormItem label={intl.get('import.endpoint')} className={styles.mixedItem}>
+        <FormItem name="platform" rules={[{ required: true, message: intl.get('formRules.platformRequired') } ]}>
+          <Select placeholder={intl.get('import.selectPlatform')}>
+            <Select.Option value="aws">AWS S3</Select.Option>
+            <Select.Option value="aliyun">OSS</Select.Option>
```
Comment (jerryliang64) [inline app/pages/Import/DatasourceList/DatasourceConfig/PlatformConfig/S3ConfigForm.tsx]: Aliyun OSS?

Comment (jerryliang64) [reply app/pages/Import/DatasourceList/DatasourceConfig/PlatformConfig/S3ConfigForm.tsx]: Tecent COS

Comment (jerryliang64) [reply app/pages/Import/DatasourceList/DatasourceConfig/PlatformConfig/S3ConfigForm.tsx]: S3 is a standard， also should consider the local s3 service

File: app/pages/Import/TaskCreate/SchemaConfig/FileMapping/FileSelectModal.tsx
Comment (jerryliang64) [inline app/pages/Import/TaskCreate/SchemaConfig/FileMapping/FileSelectModal.tsx:1]: recommend to use one component per file, easy to understand

File: server/api/studio/internal/service/datasource.go [hunk 6]
```diff
@@ -54,14 +60,14 @@ func (d *datasourceService) Add(request types.DatasourceAddRequest) (*types.Data
...
-		if err := d.testConnectionS3(c.Endpoint, c.Region, c.AccessKey, c.AccessSecret); err != nil {
+		if err := d.testConnectionS3(c.Endpoint, c.Region, c.Bucket, c.AccessKey, c.AccessSecret); err != nil {
 			return nil, err
 		}
 		secret := c.AccessSecret
 		c.AccessSecret = ""
 		cstr, err := json.Marshal(c)
 		if err != nil {
-			return nil, ecode.WithBadRequest(err, "json stringify error")
+			return nil, ecode.WithErrorMessage(ecode.ErrBadRequest, err, "json stringify error")
```
Comment (jerryliang64) [inline server/api/studio/internal/service/datasource.go]: WithBadRequest is more short and proper?

File: server/api/studio/internal/service/datasource.go [hunk 7]
```diff
@@ -250,7 +345,52 @@ func (d *datasourceService) getFileStore(dbs *db.Datasource) (filestore.FileStor
...
+		// Format: https://<bucket-name>.s3.<region>.amazonaws.com or https://s3.amazonaws.com/<bucket-name>
+		if parts[0] == "s3.amazonaws" {
+			// Format: https://s3.amazonaws.com/<bucket-name>
+			endpoint = fmt.Sprintf("https://%s", u.Host)
+			if u.Path != "" {
+				pathParts := strings.SplitN(u.Path, "/", 3)
+				bucket = pathParts[1]
+			}
+		} else {
+			// Format: https://<bucket-name>.s3.<region>.amazonaws.com
```
Comment (jerryliang64) [inline server/api/studio/internal/service/datasource.go]: can it also work for aliyun oss?

File: server/api/studio/internal/service/datasource.go [hunk 8]
```diff
@@ -282,21 +432,21 @@ func (d *datasourceService) testConnectionSFTP(host string, port int, username s
...
-		return ecode.WithBadRequest(err, "failed to dial SSH server")
+		return ecode.WithErrorMessage(ecode.ErrBadRequest, err, "failed to dial SSH server")
 	}
 	defer conn.Close()
 
 	// create an SFTP client session
 	client, err := sftp.NewClient(conn)
 	if err != nil {
-		return ecode.WithBadRequest(err, "failed to create SFTP session")
+		return ecode.WithErrorMessage(ecode.ErrBadRequest, err, "failed to create SFTP session")
```
Comment (jerryliang64) [inline server/api/studio/internal/service/datasource.go]: ecode.WithBadRquest is the short-mode of ecode.WithErrorMessage(ecode.ErrBadRequest....

File: app/interfaces/import.ts (hunk 3 above)
Comment (hetao92) [reply app/interfaces/import.ts:67]: they are the parameters of the importer, key is the file path

File: app/interfaces/import.ts (hunk 4 above)
Comment (hetao92) [reply app/interfaces/import.ts:79]: Keep the configuration definition the same as the nebula importer, but it is not currently used in the studio

File: server/api/studio/internal/service/datasource.go (hunk 8 above)
Comment (hetao92) [reply server/api/studio/internal/service/datasource.go]: image
image
its not same

File: server/api/studio/internal/service/datasource.go (hunk 7 above)
Comment (hetao92) [reply server/api/studio/internal/service/datasource.go]: I split the oss logic out from here

File: server/api/studio/internal/service/datasource.go (hunk 8 above)
Comment (jerryliang64) [reply server/api/studio/internal/service/datasource.go]: @veezhang why are they different? I previously think that the `WithBadRequest` is the shortcut of ht `WithErrorMessage(ecode.ErrBadRequest...`

File: app/interfaces/import.ts (hunk 4 above)
Comment (jerryliang64) [reply app/interfaces/import.ts:79]: If they are not used in our app, you could not pre-define them now, maybe we will never use these fields, it's little redundant.

**Review by jerryliang64 — DISMISSED** [review]
LGTM

File: app/interfaces/datasource.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  's3' = 's3',
+  'sftp' = 'sftp',
+  'local' = 'local'
+}
+export enum IS3Platform {
+  'aws' = 'aws',
+  'oss' = 'oss',
+  'customize' = 'customize',
+  'tecent' = 'cos'
+}
```
Comment (huaxiabuluo) [inline app/interfaces/datasource.ts]:
```ts
export enum DatasourceType {
  S3 = 's3',
  SFTP = 'sftp',
  Local = 'local'
}

export enum S3Platform {
  AWS = 'aws',
  OSS = 'oss',
  Tecent = 'cos',
  Customize = 'customize',
}
```

File: app/interfaces/datasource.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    bucket: string;
+    region?: string;
+  };
+  sftpConfig?: {
+    host: string;
+    port: number;
+    username: string;
+    password: string;
+  };
+}
```
Comment (huaxiabuluo) [inline app/interfaces/datasource.ts]: repetitive content between `IDatasourceAdd` and `IDatasourceItem`

File: app/pages/Import/DatasourceList/DatasourceConfig/PlatformConfig/index.tsx [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+      const flag = await updateDataSource({
+        id: data.id,
+        type: _type,
+        name: '',
+        ...values
+      });
+      setLoading(false);
+      flag && (message.success(intl.get('common.updateSuccess')), onConfirm());
+    }
+  };
```
Comment (huaxiabuluo) [inline app/pages/Import/DatasourceList/DatasourceConfig/PlatformConfig/index.tsx]:
```ts
const submit = async (values: IDatasourceItem) => {
  setLoading(true);
  const payload = { type: values.type || type, name: '', ...values };

  if (mode === 'create') {
    const flag = await addDataSource(payload);
    setLoading(false);
    flag && (message.success(intl.get('common.updateSuccess')), onConfirm());
    return;
  }

  switch (payload.type) {
    case IDatasourceType.s3: {
      const c = payload.s3Config;
      payload.s3Config = c?.accessSecret === tempPwd ? { ...c, accessSecret: undefined } : c;
      break;
    }
    case IDatasourceType.sftp: {
      const c = payload.sftpConfig;
      payload.sftpConfig = c?.password === tempPwd ? { ...c, password: undefined } : c;
      break;
    }
    default:
      break;
  }
  const flag = await updateDataSource({ id: data.id, ...payload });
  // const flag = await (() => {
  //   if (mode === 'create') {
  //     return addDataSource(payload);
  //   }

  //   switch (payload.type) {
  //     case IDatasourceType.s3: {
  //       const c = payload.s3Config;
  //       payload.s3Config = c?.accessSecret === tempPwd ? { ...c, accessSecret: undefined } : c;
  //       break;
  //     }
  //     case IDatasourceType.sftp: {
  //       const c = payload.sftpConfig;
  //       payload.sftpConfig = c?.password === tempPwd ? { ...c, password: undefined } : c;
  //       break;
  //     }
  //     default:
  //       break;
  //   }
  //   return updateDataSource({ id: data.id, ...payload });
  // })();

  setLoading(false);
  flag && (message.success(intl.get('common.updateSuccess')), onConfirm());
};
```

File: app/pages/Import/DatasourceList/RemoteList/index.tsx [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+{
+  title: 'account',
+  key: ['sftpConfig', 'username']
+},
+{
+  title: 'createTime',
+  key: 'createTime',
+  render: data => data && dayjs(data).format('YYYY-MM-DD HH:mm:ss')
+}
+];
```
Comment (huaxiabuluo) [inline app/pages/Import/DatasourceList/RemoteList/index.tsx]:
```ts
const cloudColumns: TableColumnType<IDatasourceItem>[] = [
  {
    title: <Translation>{(i) => i.get('import.ipAddress')}</Translation>,
    dataIndex: ['s3Config', 'endpoint'],
  },
  {
    title: <Translation>{(i) => i.get('import.bucketName')}</Translation>,
    dataIndex: ['s3Config', 'bucket'],
  },
  {
    title: <Translation>{(i) => i.get('import.accessKeyId')}</Translation>,
    dataIndex: ['s3Config', 'accessKey'],
  },
  {
    title: <Translation>{(i) => i.get('import.region')}</Translation>,
    dataIndex: ['s3Config', 'region'],
  },
  {
    title: <Translation>{(i) => i.get('import.createTime')}</Translation>,
    dataIndex: 'createTime',
    render: (data) => data && dayjs(data).format('YYYY-MM-DD HH:mm:ss'),
  },
];

const sftpColumns: TableColumnType<IDatasourceItem>[] = [
  {
    title: <Translation>{(i) => i.get('import.ipAddress')}</Translation>,
    dataIndex: ['sftpConfig', 'host'],
    render: (_, row) => `${row.sftpConfig.host}:${row.sftpConfig.port}`,
  },
  {
    title: <Translation>{(i) => i.get('import.account')}</Translation>,
    dataIndex: ['sftpConfig', 'username'],
  },
  {
    title: <Translation>{(i) => i.get('import.createTime')}</Translation>,
    dataIndex: 'createTime',
    render: (data) => data && dayjs(data).format('YYYY-MM-DD HH:mm:ss'),
  },
];
```

File: app/pages/Import/DatasourceList/RemoteList/index.tsx [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        title={intl.get('common.ask')}
+        okText={intl.get('common.confirm')}
+        cancelText={intl.get('common.cancel')}
+      >
+        <Button className="warningBtn">
+          <Icon type="icon-studio-btn-delete" />
+        </Button>
+      </Popconfirm>
+    </div>)
+  } as any));
```
Comment (huaxiabuluo) [inline app/pages/Import/DatasourceList/RemoteList/index.tsx]:
```ts
const tableColumns: TableColumnType<IDatasourceItem>[] = useMemo(() => {
  const columns = type === IDatasourceType.s3 ? cloudColumns : sftpColumns;
  return columns.concat({
    title: intl.get('common.operation'),
    key: 'operation',
    render: (_, item) => (
      <div className={styles.operation}>
        <Button className="primaryBtn" onClick={() => editItem(item)}>
          <Icon type="icon-studio-btn-detail" />
        </Button>
        <Popconfirm
          onConfirm={() => deleteItem(item.id)}
          title={intl.get('common.ask')}
          okText={intl.get('common.confirm')}
          cancelText={intl.get('common.cancel')}
        >
          <Button className="warningBtn">
            <Icon type="icon-studio-btn-delete" />
          </Button>
        </Popconfirm>
      </div>
    ),
  });
}, [type, intl]);
```

File: app/pages/Import/TaskCreate/SchemaConfig/FileMapping/FileSelect.tsx [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+    const newPath = `${path === '/' ? '' : path}${item.name}${item.type === 'directory' ? '/' : ''}`;
+    getDatasourceDirectory(activeId, newPath);
+  }, [path]);
+
+  const handlePathBack = useCallback(async () => {
+    if(!path || path === '/') return;
+    setState({ loading: true });
+    const _path = path.slice(0, -1).split('/');
+    _path.pop();
+    const newPath = _path.join('/').length ? _path.join('/') + '/' : '';
```
Comment (huaxiabuluo) [inline app/pages/Import/TaskCreate/SchemaConfig/FileMapping/FileSelect.tsx]:
```ts
/**
 * - `/a/b/c/?` => `/a/b/`
 * - `a/b/?` => `a`
 * - `a/?` => ``
 */
const parentPath = path.replace(/(\/|^)[^/]+\/?$/, '$1');
```

File: app/pages/Import/TaskCreate/SchemaConfig/FileMapping/FileSelect.tsx [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+        withHeader: false,
+        delimiter: ',',
+        sample: data.contents.join('\r\n'),
+        path: _path,
+        datasourceId: activeId === IDatasourceType.local ? null : activeId,
+      } as any;
+      setState({ loading: false });
+      onConfirm(item, state);
+    }
+  }, [activeItem, activeId, path]);
```
Comment (huaxiabuluo) [inline app/pages/Import/TaskCreate/SchemaConfig/FileMapping/FileSelect.tsx]: early return, unnecessary `else`

**Review by jerryliang64 — APPROVED** [review]
LGTM

**Pull request merged**